agent/src/config/config.cpp

```diff
@@ -17,205 +17,189 @@
+        UniqueKey config = dep.first;
+        std::string group = std::get<0>(config);
+        std::string key = std::get<1>(config);
 
-    /* Checking for mandatory config group is only relevant for the main config source
-     */
-    for (std::string group : mandatory.groups()) {
-        log_debug() << "Checking for mandatory config group: \"" << group << "\"";
-        if (!m_config->has_group(group)) {
-            log_error() << "Missing mandatory config group: \"" + group + "\"";
-            throw ConfigMandatoryError("Missing mandatory config group: \"" + group + "\"");
+        // Check if the dependee exists in the configs we got from the sources
+        bool found = (std::find(m_allConfigs.begin(), m_allConfigs.end(), config) != m_allConfigs.end());
+
+        // If a dependee was present among the configs, all its dependencies needs to be present as well
+        if (found) {
+            if (allDepsSatisfied(dep.second)) {
+                log_debug() << "All dependencies found for config " << group << "::" << key;
+            } else {
+                // It is a fatal error if one or more dependencies were not found
+                std::string message = "One or more dependencies were not met for config " + group + "::" + key;
+                log_error() << message;
+                throw ConfigDependencyError(message);
+            }
         }
     }
+}
 
-    /* Check that each mandatory config is present in one of the sources using the normal
-     * retrieval methods in order to preserve the priority order.
-     */
-    for (std::tuple<std::string, std::string, ConfigType> config : mandatory.configs()) {
-        std::string group = std::get<0>(config);
-        std::string key = std::get<1>(config);
-        ConfigType type = std::get<2>(config);
-
-        log_debug() << "Checking for presence of mandatory config: \"" << key << "\"";
-        /* TODO: The below only checks for presence, it should also add the values to something
-         *       that can be used for retrieval later. */
-        try {
-            switch (type) {
-                case ConfigType::String:
-                    getStringValue(group, key);
-                    break;
-                case ConfigType::Integer:
-                    getIntegerValue(group, key);
-                    break;
-                case ConfigType::Boolean:
-                    getBooleanValue(group, key);
-                    break;
-                default:
-                    break;
-            }
-        } catch (ConfigError &error) {
-            log_error() << "Missing mandatory config: \"" << key << "\"";
-            throw ConfigMandatoryError("Missing mandatory config: \"" + key + "\"");
+bool Config::allDepsSatisfied(const std::vector<UniqueKey> &dependencies) const
+{
+    for (const UniqueKey &dependency : dependencies) {
+        bool found = (std::find(m_allConfigs.begin(), m_allConfigs.end(), dependency) != m_allConfigs.end());
```

Comment:
I think it's more readable actually. But I don't mind changing it, it's not strictly needed i suppose.